In [1]:
from pydataset import data
import pandas as pd
import numpy as np

In [2]:
# data('mpg', show_doc=True) # view the documentation for the dataset
mpg = data('mpg') # load the dataset and store it in a variable


In [3]:
np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})

type(df)

pandas.core.frame.DataFrame

## Q1
Q1a

In [4]:
df = df.assign(passing_english=df.english >= 70)
df

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
11     Alan    92       62       72            False

Q1b

In [5]:
df.sort_values(by=['passing_english'])
#duplicates are together and ordered by their key number

name  math  english  reading  passing_english
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True

Q1c

In [6]:
df.sort_values(by=['passing_english','name'])

name  math  english  reading  passing_english
11     Alan    92       62       72            False
8    Albert    92       62       87            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True
1      Jane    88       79       67             True
5      John    79       76       93             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
2     Suzie    94       74       95             True

Q1d

In [7]:
df.sort_values(by=['passing_english','english'])

name  math  english  reading  passing_english
8    Albert    92       62       87            False
11     Alan    92       62       72            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
2     Suzie    94       74       95             True
5      John    79       76       93             True
1      Jane    88       79       67             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True

Q1e

In [8]:
df = df.assign(gpa=(df.math + df.english + df.reading)/3)
df

name  math  english  reading  passing_english        gpa
0     Sally    62       85       80             True  75.666667
1      Jane    88       79       67             True  78.000000
2     Suzie    94       74       95             True  87.666667
3     Billy    98       96       88             True  94.000000
4       Ada    77       92       98             True  89.000000
5      John    79       76       93             True  82.666667
6    Thomas    82       64       81            False  75.666667
7     Marie    93       63       90            False  82.000000
8    Albert    92       62       87            False  80.333333
9   Richard    69       80       94             True  81.000000
10    Isaac    92       99       93             True  94.666667
11     Alan    92       62       72            False  75.333333

## Q2

In [9]:
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

In [10]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


In [11]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

In [12]:
mpg = mpg.rename(columns = {'cty':'city'})
mpg

manufacturer   model  displ  year  cyl       trans drv  city  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16   26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

In [13]:
mpg = mpg.rename(columns = {'hwy':'highway'})
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

In [14]:
mpg.sort_values(by=['city', 'highway'])

manufacturer                model  displ  year  cyl       trans drv  city  \
55         dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
60         dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
66         dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
70         dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
127         jeep   grand cherokee 4wd    4.7  2008    8    auto(l5)   4     9   
..           ...                  ...    ...   ...  ...         ...  ..   ...   
100        honda                civic    1.6  1999    4  manual(m5)   f    28   
197       toyota              corolla    1.8  2008    4  manual(m5)   f    28   
223   volkswagen           new beetle    1.9  1999    4    auto(l4)   f    29   
213   volkswagen                jetta    1.9  1999    4  manual(m5)   f    33   
222   volkswagen           new beetle    1.9  1999    4  manual(m5)   f    35   

     highway fl       class  
55        12  e      pickup  
60        12  e         suv  
66        12  e      pickup  
70        12  e      pickup  
127       12  e         suv  
..       ... ..         ...  
100       33  r  subcompact  
197       37  r     compact  
223       41  d  subcompact  
213       44  d     compact  
222       44  d  subcompact  

[234 rows x 11 columns]

In [15]:
mpg.assign(bettercity= mpg.city>mpg.highway)

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  bettercity  
1    compact       False  
2    compact       False  
3    compact       False  
4    compact       False  
5    compact       False  
..       ...         ...  
230  midsize       False  
231  midsize       False  
232  midsize       False  
233  midsize       False  
234  midsize       False  

[234 rows x 12 columns]

No car has better city mileage than highway mileage.

In [16]:
mpg = mpg.assign(mileage_difference=mpg.highway-mpg.city)
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

In [17]:
mpg.sort_values(by='mileage_difference', ascending = False)

manufacturer              model  displ  year  cyl       trans drv  city  \
107        honda              civic    1.8  2008    4    auto(l5)   f    24   
223   volkswagen         new beetle    1.9  1999    4    auto(l4)   f    29   
1           audi                 a4    1.8  1999    4    auto(l5)   f    18   
229   volkswagen             passat    1.8  1999    4    auto(l5)   f    18   
36     chevrolet             malibu    3.5  2008    6    auto(l4)   f    18   
..           ...                ...    ...   ...  ...         ...  ..   ...   
80          ford       explorer 4wd    4.0  1999    6    auto(l5)   4    14   
138      mercury    mountaineer 4wd    4.0  1999    6    auto(l5)   4    14   
177       toyota        4runner 4wd    3.4  1999    6  manual(m5)   4    15   
152       nissan     pathfinder 4wd    3.3  1999    6  manual(m5)   4    15   
204       toyota  toyota tacoma 4wd    3.4  1999    6  manual(m5)   4    15   

     highway fl       class  mileage_difference  
107       36  c  subcompact                  12  
223       41  d  subcompact                  12  
1         29  p     compact                  11  
229       29  p     midsize                  11  
36        29  r     midsize                  11  
..       ... ..         ...                 ...  
80        17  r         suv                   3  
138       17  r         suv                   3  
177       17  r         suv                   2  
152       17  r         suv                   2  
204       17  r      pickup                   2  

[234 rows x 12 columns]

Honda Civics and Volkswagen New Beetles have the highest mileage difference.

In [18]:
compactdf = mpg[mpg['class'] == 'compact']

In [19]:
compactdf.sort_values(by='highway')

manufacturer         model  displ  year  cyl       trans drv  city  \
220   volkswagen         jetta    2.8  1999    6    auto(l4)   f    16   
221   volkswagen         jetta    2.8  1999    6  manual(m5)   f    17   
212   volkswagen           gti    2.8  1999    6  manual(m5)   f    17   
172       subaru   impreza awd    2.5  2008    4  manual(m5)   4    19   
170       subaru   impreza awd    2.5  2008    4    auto(s4)   4    20   
15          audi    a4 quattro    3.1  2008    6  manual(m6)   4    15   
14          audi    a4 quattro    3.1  2008    6    auto(s6)   4    17   
13          audi    a4 quattro    2.8  1999    6  manual(m5)   4    17   
9           audi    a4 quattro    1.8  1999    4    auto(l5)   4    16   
12          audi    a4 quattro    2.8  1999    6    auto(l5)   4    15   
209   volkswagen           gti    2.0  1999    4    auto(l4)   f    19   
215   volkswagen         jetta    2.0  1999    4    auto(l4)   f    19   
8           audi    a4 quattro    1.8  1999    4  manual(m5)   4    18   
191       toyota  camry solara    3.0  1999    6    auto(l4)   f    18   
6           audi            a4    2.8  1999    6  manual(m5)   f    18   
5           audi            a4    2.8  1999    6    auto(l5)   f    16   
192       toyota  camry solara    3.0  1999    6  manual(m5)   f    18   
187       toyota  camry solara    2.2  1999    4    auto(l4)   f    21   
173       subaru   impreza awd    2.5  2008    4  manual(m5)   4    20   
193       toyota  camry solara    3.3  2008    6    auto(s5)   f    18   
143       nissan        altima    2.4  1999    4    auto(l4)   f    19   
11          audi    a4 quattro    2.0  2008    4    auto(s6)   4    19   
7           audi            a4    3.1  2008    6    auto(av)   f    18   
171       subaru   impreza awd    2.5  2008    4    auto(s4)   4    20   
10          audi    a4 quattro    2.0  2008    4  manual(m6)   4    20   
216   volkswagen         jetta    2.0  2008    4    auto(s6)   f    22   
214   volkswagen         jetta    2.0  1999    4  manual(m5)   f    21   
217   volkswagen         jetta    2.0  2008    4  manual(m6)   f    21   
218   volkswagen         jetta    2.5  2008    5    auto(s6)   f    21   
219   volkswagen         jetta    2.5  2008    5  manual(m5)   f    21   
211   volkswagen           gti    2.0  2008    4    auto(s6)   f    22   
210   volkswagen           gti    2.0  2008    4  manual(m6)   f    21   
208   volkswagen           gti    2.0  1999    4  manual(m5)   f    21   
1           audi            a4    1.8  1999    4    auto(l5)   f    18   
188       toyota  camry solara    2.2  1999    4  manual(m5)   f    21   
142       nissan        altima    2.4  1999    4  manual(m5)   f    21   
2           audi            a4    1.8  1999    4  manual(m5)   f    21   
194       toyota       corolla    1.8  1999    4    auto(l3)   f    24   
4           audi            a4    2.0  2008    4    auto(av)   f    21   
190       toyota  camry solara    2.4  2008    4    auto(s5)   f    22   
3           audi            a4    2.0  2008    4  manual(m6)   f    20   
189       toyota  camry solara    2.4  2008    4  manual(m5)   f    21   
195       toyota       corolla    1.8  1999    4    auto(l4)   f    24   
196       toyota       corolla    1.8  1999    4  manual(m5)   f    26   
198       toyota       corolla    1.8  2008    4    auto(l4)   f    26   
197       toyota       corolla    1.8  2008    4  manual(m5)   f    28   
213   volkswagen         jetta    1.9  1999    4  manual(m5)   f    33   

     highway fl    class  mileage_difference  
220       23  r  compact                   7  
221       24  r  compact                   7  
212       24  r  compact                   7  
172       25  p  compact                   6  
170       25  p  compact                   5  
15        25  p  compact                  10  
14        25  p  compact                   8  
13        25  p  compact                   8  
9         25  p  compact    

2 variations of the Volkswagen Jetta are owners of the best and worst highway mileage

In [20]:
mpg = mpg.assign(average_mileage=(mpg.city+mpg.highway)/2)

In [21]:
dodgedf=mpg[mpg['manufacturer']=='dodge']

In [22]:
dodgedf

manufacturer                model  displ  year  cyl       trans drv  city  \
38        dodge          caravan 2wd    2.4  1999    4    auto(l3)   f    18   
39        dodge          caravan 2wd    3.0  1999    6    auto(l4)   f    17   
40        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
41        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
42        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
43        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
44        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    11   
45        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
46        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
47        dodge          caravan 2wd    3.8  2008    6    auto(l6)   f    16   
48        dodge          caravan 2wd    4.0  2008    6    auto(l6)   f    16   
49        dodge    dakota pickup 4wd    3.7  2008    6  manual(m6)   4    15   
50        dodge    dakota pickup 4wd    3.7  2008    6    auto(l4)   4    14   
51        dodge    dakota pickup 4wd    3.9  1999    6    auto(l4)   4    13   
52        dodge    dakota pickup 4wd    3.9  1999    6  manual(m5)   4    14   
53        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
54        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
55        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
56        dodge    dakota pickup 4wd    5.2  1999    8  manual(m5)   4    11   
57        dodge    dakota pickup 4wd    5.2  1999    8    auto(l4)   4    11   
58        dodge          durango 4wd    3.9  1999    6    auto(l4)   4    13   
59        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
60        dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
61        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
62        dodge          durango 4wd    5.2  1999    8    auto(l4)   4    11   
63        dodge          durango 4wd    5.7  2008    8    auto(l5)   4    13   
64        dodge          durango 4wd    5.9  1999    8    auto(l4)   4    11   
65        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
66        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
67        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
68        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
69        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
70        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
71        dodge  ram 1500 pickup 4wd    5.2  1999    8    auto(l4)   4    11   
72        dodge  ram 1500 pickup 4wd    5.2  1999    8  manual(m5)   4    11   
73        dodge  ram 1500 pickup 4wd    5.7  2008    8    auto(l5)   4    13   
74        dodge  ram 1500 pickup 4wd    5.9  1999    8    auto(l4)   4    11   

    highway fl    class  mileage_difference  average_mileage  
38       24  r  minivan                   6             21.0  
39       24  r  minivan                   7             20.5  
40       22  r  minivan                   6             19.0  
41       22  r  minivan                   6             19.0  
42       24  r  minivan                   7             20.5  
43       24  r  minivan                   7             20.5  
44       17  e  minivan                   6             14.0  
45       22  r  minivan                   7             18.5  
46       21  r  minivan                   6             18.0  
47       23  r  minivan                   7             19.5  
48       23  r  minivan                   7             19.5  
49       19  r   pickup                   4             17.0  
50       18  r   pickup                   4             16.0  
51       17  r   pickup                   4             15.0  
52       17  r   

Dodge Caravan 2wd has the best average mileage
Dodge Ram 1500 pickup 4wd has the worst average mileage

## Q3

In [23]:
Mammals = data('Mammals')

In [26]:
Mammals.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 107 entries, 1 to 107
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   weight    107 non-null    float64
 1   speed     107 non-null    float64
 2   hoppers   107 non-null    bool   
 3   specials  107 non-null    bool   
dtypes: bool(2), float64(2)
memory usage: 2.7 KB


107 rows, 4 columns

In [28]:
Mammals.dtypes

weight      float64
speed       float64
hoppers        bool
specials       bool
dtype: object

In [29]:
Mammals.describe()

weight       speed
count   107.000000  107.000000
mean    278.688178   46.208411
std     839.608269   26.716778
min       0.016000    1.600000
25%       1.700000   22.500000
50%      34.000000   48.000000
75%     142.500000   65.000000
max    6000.000000  110.000000

In [31]:
Mammals.sort_values(by='speed', ascending = False)

weight  speed  hoppers  specials
53     55.000  110.0    False     False
39     37.000  105.0    False     False
35     50.000  100.0    False     False
41     34.000   97.0    False     False
42     30.000   97.0    False     False
19    300.000   90.0    False     False
34     62.000   81.0    False     False
45     20.000   81.0    False     False
25    120.000   80.0    False     False
22    170.000   80.0    False     False
20    250.000   80.0    False     False
40     35.000   80.0    False     False
43     30.000   80.0    False     False
63      6.000   72.0    False     False
18    300.000   72.0    False     False
98      4.000   72.0     True     False
97      4.400   72.0     True     False
12    900.000   70.0    False     False
57     20.000   70.0    False     False
8     260.000   70.0    False     False
6     350.000   70.0    False     False
24    130.000   70.0    False     False
5     400.000   70.0    False     False
56     25.000   67.0    False     False
30     80.000   65.0    False     False
58     16.000   65.0    False     False
51     65.000   65.0    False     False
104    50.000   65.0    False     False
28    100.000   64.0    False     False
64      5.000   64.0    False     False
100     2.000   64.0     True     False
55     40.000   64.0    False     False
96      4.600   64.0     True     False
7     300.000   64.0    False     False
26    120.000   61.0    False     False
36     50.000   60.0    False     False
11   1000.000   60.0    False     False
33     65.000   60.0    False     False
52     60.000   60.0    False     False
62      7.000   60.0    False     False
49    150.000   59.0    False     False
15    750.000   57.0    False     False
17    450.000   56.0    False     False
101     1.900   56.0     True     False
13    900.000   56.0    False     False
21    250.000   56.0    False     False
61     10.000   56.0    False     False
99      3.500   56.0     True     False
48    230.000   56.0    False     False
31     72.000   56.0    False     False
29     85.000   55.0    False     False
54     45.000   50.0    False     False
102     1.500   50.0     True     False
50    135.000   48.0    False     False
47    300.000   48.0    False     False
23    150.000   48.0    False     False
38     50.000   47.0    False     False
44     30.000   45.0    False     False
32     70.000   45.0    False     False
4    1400.000   45.0    False     False
46    400.000   40.0    False     False
37     50.000   40.0    False     False
68     70.000   40.0    False      True
103     1.500   40.0     True     False
9     250.000   40.0    False     False
69     13.000   37.0    False      True
72      0.600   36.0    False     False
1    6000.000   35.0    False     False
27    110.000   33.0    False     False
67    127.000   32.0    False     False
16    500.000   32.0    False     False
85      0.035   32.0     True     False
60     11.000   30.0    False      True
14    800.000   29.0    False     False
65      5.000   27.0    False      True
74      0.550   27.0    False     False
2    4000.000   26.0    False     False
3    3000.000   25.0    False     False
10   3800.000   25.0    False      True
59     12.000   24.0    False      True
82      0.056   21.0     True     False
76      0.400   20.0    False     False
73      0.600   20.0    False     False
75      0.500   18.0    False     False
81      0.100   17.0    False     False
84      0.045   16.0    False     False
71      4.000   16.0    False     False
66      3.000   16.0    False      True
79      0.220   15.0    False     False
86      0.035   14.0     True     False
92      0.016   13.0    False     False
77      0.300   13.0    False     False
106     0.024   13.0    False     False
89      0.025   11.0    False     False
83      0.050   11.0    False     False
78      0.250    9.7    False     False
88      0.030    9.1    False     False
80      0.110    9.0    False     False
91      0.018    8.9    False     False
90     

Weight of fastest animal is 55.00

In [37]:
IsSpecial = Mammals[Mammals['specials'] == True].count()

In [38]:
IsSpecial

weight      10
speed       10
hoppers     10
specials    10
dtype: int64

In [55]:
IsSpecial/Mammals.count()

weight      0.093458
speed       0.093458
hoppers     0.093458
specials    0.093458
dtype: float64

9.35% specials

In [49]:
medianspeeddf = Mammals[Mammals['speed']>Mammals['speed'].median()]
medianspeeddf

weight  speed  hoppers  specials
5     400.0   70.0    False     False
6     350.0   70.0    False     False
7     300.0   64.0    False     False
8     260.0   70.0    False     False
11   1000.0   60.0    False     False
12    900.0   70.0    False     False
13    900.0   56.0    False     False
15    750.0   57.0    False     False
17    450.0   56.0    False     False
18    300.0   72.0    False     False
19    300.0   90.0    False     False
20    250.0   80.0    False     False
21    250.0   56.0    False     False
22    170.0   80.0    False     False
24    130.0   70.0    False     False
25    120.0   80.0    False     False
26    120.0   61.0    False     False
28    100.0   64.0    False     False
29     85.0   55.0    False     False
30     80.0   65.0    False     False
31     72.0   56.0    False     False
33     65.0   60.0    False     False
34     62.0   81.0    False     False
35     50.0  100.0    False     False
36     50.0   60.0    False     False
39     37.0  105.0    False     False
40     35.0   80.0    False     False
41     34.0   97.0    False     False
42     30.0   97.0    False     False
43     30.0   80.0    False     False
45     20.0   81.0    False     False
48    230.0   56.0    False     False
49    150.0   59.0    False     False
51     65.0   65.0    False     False
52     60.0   60.0    False     False
53     55.0  110.0    False     False
54     45.0   50.0    False     False
55     40.0   64.0    False     False
56     25.0   67.0    False     False
57     20.0   70.0    False     False
58     16.0   65.0    False     False
61     10.0   56.0    False     False
62      7.0   60.0    False     False
63      6.0   72.0    False     False
64      5.0   64.0    False     False
96      4.6   64.0     True     False
97      4.4   72.0     True     False
98      4.0   72.0     True     False
99      3.5   56.0     True     False
100     2.0   64.0     True     False
101     1.9   56.0     True     False
102     1.5   50.0     True     False
104    50.0   65.0    False     False

In [51]:
medianspeeddf[medianspeeddf['hoppers'] == True].count()

weight      7
speed       7
hoppers     7
specials    7
dtype: int64

7 hoppers above the median speed

In [54]:
medianspeeddf[medianspeeddf['hoppers'] == True].count()/medianspeeddf.count()

weight      0.132075
speed       0.132075
hoppers     0.132075
specials    0.132075
dtype: float64

13.21% of animals above the median speed are hoppers